### ISSUE TITLE
module vault-elb is not idempotent

### ISSUE THREAD

**Issue by kostyrev** [hashicorp/terraform-aws-vault#19]
I'm creating ELB for vault with this:
```
module "vault_elb" {
  source                      = "git::git@github.com:hashicorp/terraform-aws-vault.git//modules/vault-elb"
  internal                    = true
  name                        = "${var.country}-${var.stage}-vault-elb"
  vpc_id                      = "${module.network-getter.vpc_id}"
  subnet_ids                  = "${module.network-getter.private_subnet_ids}"
  allowed_inbound_cidr_blocks = ["${data.aws_vpc.main.cidr_block}"]
}
```
everything applies with no problem. Thank you guys!
But after invoking `terraform plan` once again I see the following:
```
Terraform will perform the following actions:
  ~ module.vault_elb.aws_elb.vault
      availability_zones.#:          "3" => "0"
      availability_zones.1305112097: "us-east-1b" => ""
      availability_zones.3569565595: "us-east-1a" => ""
      availability_zones.986537655:  "us-east-1c" => ""


Plan: 0 to add, 1 to change, 0 to destroy.
```
```
$ terraform version
Terraform v0.10.7
```

**Comment by brikis98** [comment]
99% of the time, these sorts of issues are bugs in Terraform itself. The vault-elb module is setting `availability_zones` to an empty list by default; my guess is that after creating the ELB, AWS fills in the AZ data automatically, so next time you run `plan`, Terraform figures that the empty AZ list means "remove all AZs". 

Perhaps file a bug in the [AWS provider](https://github.com/terraform-providers/terraform-provider-aws)?

**Comment by kchugalinskiy** [comment]
```
data "aws_subnet" "private" {
  id = "${var.private_subnets}"

  count = "${length(var.private_subnets)}"
}

module "vault_elb" {
....blah-blah-blah....
  availability_zones          = "${data.aws_subnet.private.*.availability_zone}"
}
```
@kostyrev this may be a workaround for your problem :-)

@brikis98 maybe this should be used as a default value for `availability_zones` argument. This datasource is available since version 0.9.5 which is minimal in your case. Proof: https://github.com/hashicorp/terraform/blob/v0.9.5/website/source/docs/providers/aws/d/subnet.html.markdown

Guys, what do you think?

**Comment by brikis98** [comment]
@kchugalinskiy Nice workaround. I'd welcome a PR for that!

**Comment by kchugalinskiy** [comment]
@brikis98 and here we go: https://github.com/hashicorp/terraform-aws-vault/pull/20

**Comment by brikis98** [comment]
Thx @kchugalinskiy! Fixed by #20. Available in https://github.com/hashicorp/terraform-aws-vault/releases/tag/v0.0.7.

**Comment by brikis98** [comment]
Unfortunately, I had to revert the fix in #20 due to the issue explained in #21. That means that for now, the elb module will always show a diff when you run `plan`. It's harmless and can be safely ignored. I've filed a bug with the Terraform AWS provider about this: https://github.com/terraform-providers/terraform-provider-aws/issues/2156.

**Comment by holybit** [comment]
I understand the issue, but can we solve this faster by not worrying about what Hashicorp might do with [issue](https://github.com/terraform-providers/terraform-provider-aws/issues/2156)?

`aws_elb` states clearly in the [docs](https://www.terraform.io/docs/providers/aws/r/elb.html):

> Exactly one of availability_zones or subnets must be specified: this determines if the ELB exists in a VPC or in EC2-classic.

Yes, Hashicorp could enhance `aws_elb` resource to accept a null value for one or the other but not both arguments.

But, is anyone seriously using Vault in EC2-classic VPC setup? It's been deprecated for all new AWS accounts since 2013-12-04. I would say the answer is close enough to no that removing the `availability_zones` would be safe and expedient to resolving this issue. Things are so bad right now you can't eve get this module to work as it just bars up a big fat red error.

**Comment by brikis98** [comment]
> Things are so bad right now you can't eve get this module to work as it just barfs up a red error.

Wait, what error were you seeing? I'm aware of a diff showing up every time, but you could safely ignore it. Are you seeing something else now?

**Comment by patoarvizu** [comment]
The problem is that the first `terraform apply` works, but subsequent `apply`s show the `availability_zones.#:          "3" => "0"` diff and it fails because of a restriction in ELBs where you can't remove all AZ at the same time (i.e. you can't be with 0 AZs). This is a restriction on the actual Amazon resource, not on the AWS Terraform provider.

I think I agree with @holybit, can we consider removing support for EC2-classic?

**Comment by brikis98** [comment]
> it fails because of a restriction in ELBs where you can't remove all AZ at the same time

Hmm, I wonder if something changed? I don't think this would cause an error before. Simply running `apply` despite the log output used to work just fine, making no actual changes. 

At any rate, if this is the case, I'd welcome a PR to remove the `availability_zones` param.

**Comment by patoarvizu** [comment]
I'm submitting two alternative PRs for this shortly. I'll let the maintainers decide which approach they prefer :)

**Comment by brikis98** [comment]
This should now be fixed by #44. Thx @patoarvizu!
